pages/v4/Coroutines.md

```diff
@@ -0,0 +1,184 @@
+	}
+}
+```
+
+Which is fine, and works well for you.
+But now, let's assume you want the transition to be smooth; one way would be to move 1 unit per frame (let's ignore the time it would take or the fact that it would be dependent on the frame rate)
+
+```csharp
+private Vector2? movementTarget = null;
+private Vector2 movementDirection = Vector2.Zero;
+
+public void OnUpdate()
+{
+	if(movementTarget != null)
+	{
+		if((this.movementTarget.Value - this.GameObj.Transform.Pos.Xy).LengthSquared >= this.movementDirection.LengthSquared)
+		{
+			this.GameObj.Transform.Pos.MoveBy(movementDirection);
+		}
+		else
+		{
+			this.GameObj.Transform.Pos.MoveTo(this.movementTarget.Value);
+			this.movementTarget = this.Game;
+		}
+	}
+	else if (DualityApp.Keyboard.KeyHit(Key.Space))
+	{
+		this.movementTarget = this.GameObj.Transform.Pos.Xy + new Vector2(200, 0);
+		this.movementDirection = (this.movementTarget.Value - this.GameObj.Transform.Pos.Xy).Normalized;
+	}
+}
+```
+
+Again, this works, but the OnUpdate method is starting to appear cluttered with unnecessary code; wouldn't it be great if the part of code in charge of managing the movement was in another part?
+Sure, you could refactor the code to extract the method and have a `UpdateColor` or similar being called at the end or beginning of `OnUpdate`; good, but we can still do better. We can make a Coroutine out of it.
+
+A Coroutine is implemented as a method that returns an `IEnumerable<WaitUntil>` collection, and whose body is separated in execution sections by `yield return` statements. At instantiation, and at each Update loop, the method is executed from its current point until the next `yield`, and so its total execution can be spread out over a number of frames.
+
+To get back to our previous example, we could have our Coroutine as follows:
+
+```csharp
+private IEnumerable<WaitUntil> MoveTo(Vector2 target)
+{
+	Vector2 movement = (target - this.GameObj.Transform.Pos.Xy).Normalized;
+	
+	while((target - this.GameObj.Transform.Pos.Xy).LengthSquared >= movement.LengthSquared)
+	{
+		this.GameObj.Transform.Pos.MoveBy(movement);
+		yield return WaitUntil.NextFrame;
+	}
+	
+	this.GameObj.Transform.Pos.MoveTo(target);
+}
+```
+
+Let's see it more in detail:
+
+```csharp
+private IEnumerable<WaitUntil> MoveTo(Vector2 target)
+```
+
+this is our Coroutine definition; note that, being a method like all others, it can have parameters and it can have any kind of access modifier, as long as it is accessible by its caller.
+
+```csharp
+Vector2 movement = (target - this.GameObj.Transform.Pos.Xy).Normalized;
+```
+
+We don't need anymore to keep a private class variable; now we can just keep our movement data inside the Coroutine and declutter the rest of the class from things that are used only in specific cases
+
+```csharp
+while((target - this.GameObj.Transform.Pos.Xy).LengthSquared >= movement.LengthSquared)
+{
+	this.GameObj.Transform.Pos.MoveBy(movement);
+```
```

Comment:
Missing }

Reply:
ah, here is just because I just want to focus on those 3 lines from the whole method :)

Reply:
yeah but pasting this will cause the code editor to scream lots of errors at you now :P

Reply:
but you should past the whole method 5 paragraph before 😄

Reply:
Hmm you could also omit the other brace as thats still valid C# :P